crowbar_framework/app/models/api/upgrade.rb

```diff
@@ -1442,6 +1442,15 @@ def non_disruptive_upgrade_compute_nodes(names)
           ::Node.find_node_by_name_or_alias(name)
         end
 
+        # remove nodes which were already upgraded (it might leave an empty list)
+        compute_nodes.reject!(&:upgraded?)
+        if compute_nodes.empty?
+          Rails.logger.info(
+            "All requested compute nodes are already upgraded."
+          )
+          return
+        end
+
```

Comment:
So, I think I've changed my mind on this :-)

I think the `reject` call above is correct and we should keep it to (silently) filter out upgraded nodes. However if all provided nodes are already upgraded, we should raise an error. Preferably in controller already, as you suggested earlier. This place https://github.com/crowbar/crowbar-core/blob/master/crowbar_framework/app/controllers/api/upgrade_controller.rb#L189 seems good for it as we are already checking all nodes data.

Reply:
done